### PR TITLE
Fix #211  Optimize reduce-scatter for Q matrix

### LINKED ISSUE #211: Optimize reduce-scatter for Q matrix
Reduce-scatter is used for synchronizing Q matrix, and it takes a lot of time for big multi-node simulations, see e.g. plots in #209.

# Current algorithm

(implemented in #142, to be merged to master soon)

See  https://github.com/davidsd/sdpb/blob/00e05c3c74e6a5cb3858a765994acba618b63167/src/sdp_solve/SDP_Solver/run/bigint_syrk/BigInt_Shared_Memory_Syrk_Context/reduce_scatter_DistMatrix.hxx

We have one `DistMatrix` per node as an input. `reduce_scatter()` calculates their sum and writes the result to a global `DistMatrix` output.

Current implementation uses ring algorithm for all ranks. Each rank performs `total_ranks - 1` MPI sends to the next ranks. Number of elements in each send/recv message [equals](https://github.com/davidsd/sdpb/blob/00e05c3c74e6a5cb3858a765994acba618b63167/src/sdp_solve/SDP_Solver/run/bigint_syrk/BigInt_Shared_Memory_Syrk_Context/reduce_scatter_DistMatrix.hxx#L83) to the number of output matrix elements owned by the corresponding rank, which can be estimated as `#(Q)/total_ranks`.

As a result, each rank sends `~ #(Q)` elements in total to the next rank.
Total amount of data sent is `~ #(Q) * total_ranks` bigfloats.
`~ #(Q) * num_nodes` of them are sent between nodes (from the last rank of a node to the first rank of the next node), and the remaining elements are sent within a node (which is much faster).

Each rank send_receive buffer:
https://github.com/davidsd/sdpb/blob/00e05c3c74e6a5cb3858a765994acba618b63167/src/sdp_solve/SDP_Solver/run/bigint_syrk/BigInt_Shared_Memory_Syrk_Context/reduce_scatter_DistMatrix.hxx#L86-L90

Size of the buffer is `~ 2 * #(Q) / total_ranks` bigfloats for each rank, i.e. `~ 2 * #(Q)` in total.

# Possible improvement

In the optimal case, `~ #(Q) * (num_nodes - 1)` data will be sent between the nodes, and no data will be sent within the node.

How can we achieve that?

Let's group together all ranks with the same rank_in_node: e.g. for three 128-core nodes the groups will be `(0, 128, 256)`, `(1, 129, 257)` etc.
Each element of the output matrix is owned by some rank belonging to some group.
Suppose that some element Q_ij is owned by rank 129.
Then the ranks 1 and 257 should calculate node contribution to Q_ij and send it to rank 129. Rank 129 will calculate its own contribution, add it to the received data, and write the result to Q_ij.
In other words, we perform MPI reduce of `input(i, j)` to `output(i, j)` over the ranks (1, 129, 257) to rank 129.

Instead of sending individual matrix elements, we may send an array of all elements owned by a given rank. This should work faster, but sending larger buffers will increase memory consumption - naive implementation (buffering all sends and receives simultaneously) would require buffer size of `~ 2 * #(Q) * (num_nodes - 1)` in total across all ranks.

P.S. Thanks to PR #209, we can always reduce memory by performing reduce-scatter only for submatrices of Q (but this would cause slowdown in other parts of bigint_syrk_blas).

### PR DESCRIPTION
**Pull request by vasdommes**
- Fix #211

**TL;DR** In our benchmark (see below), reduce-scatter became **~50x** faster. Thanks to that, SDPB became **~2x** faster on big problems, and scales much better for large number of nodes.

Old SDPB version uses ring algorithm among all ranks to accumulate local contributions to the Q matrix and write them to the global DistMatrix.
Ring algorithm consists of `num_ranks - 1`  iterations, and each rank was sending `~ #(Q) / num_ranks` matrix elements to another rank.

That made sense before we started using shared memory window for calculating Q, see #142.
Now all ranks on a node have access to the shared memory window containing residues of Q_n (contribution to Q from node n) modulo all primes.

Therefore, no communication inside a node is required, and all we have to do is to reduce Q_n from all nodes into global Q.
If a node owns some element `Q[i, j]`, then all other nodes should send their `Q_n[i, j]` to that node. In addition, the node restores its own contribution from residues.

If each node has one rank, then the implemented algorithm works as follows:
```
for offset = 1..num_nodes-1:
  - Each node (n) restores elements Q_n[i,j] for all [i,j] owned by node (n+offset) from residues,
    and sends them to node (n+offset). Implemented via MPI_Sendrecv.
  - Each node updates its own Q[i,j] with data received from node (n-offset).
```

Since each node usually has multiple ranks, we have to distribute the job between them. We use the following scheme:
If rank r owns the given element Q[i,j], then it will receive contributions from other nodes from the processes having the same rank within a node (we use the fact that all nodes have the same number of ranks).
For example, for 3 nodes with 128 cores, rank 0 will communicate only with ranks 128 and 256, rank 1 - with ranks 129 and 257, and so on. As a result, `MPI_Sendrecv` is called on communicators `(0,128,256)`, `(1,129,257)` etc.

Each rank has send and receive buffers of size `~ #(Q) / num_ranks`, and performs `(num_nodes - 1)` send/recv operations.

### CODE REVIEW

**Comment by vasdommes** [comment]
Benchmarks on Expanse HPC for nmax=18 stress-tensors-3d:
Q is 5485x5485 matrix, precision=1024.

We see that the old reduce-scatter algorithm (first plot) took significant amount of time (**5-10 minutes** per solver iteration). Time grows linearly with the number of nodes and does not allow to speed up SDPB by increasing number of nodes.

With the new algorithm (second plot), reduce-scatter works really fast (~**10 seconds** per solver iteration!), and overall SDPB performance scales very well even for 10+ nodes.

Overall, SDPB is now ~2x faster on 7 nodes and 3.5x faster for 12 nodes.


image
image

**Comment by vasdommes** [comment]
Note also that communication itself (mpi_sendrecv) is fast, most of the time is spent on serializing/deserializing BigFloat numbers.
image
